### PR TITLE
[WIP][Core][Refactor] move vllm/model_executor/parallel_utils into vllm/distributed and vllm/device_communicators

### PR DESCRIPTION
**Pull request by youkaichao**
I don't place `vllm/device_communicators` under `vllm/distributed/device_communicators`, because device communicators can also be used in a single process.

Main idea of the change:

`vllm/model_executor/parallel_utils/parallel_state.py` and `vllm/model_executor/parallel_utils/communication_op.py` --> `vllm/distributed.py`

the rest: --> `vllm/device_communicators`

Seek for suggestions @zhuohan123 , let's discuss the structure of refactor first, then fix test and linter.

### CODE REVIEW

File: vllm/utils.py [hunk 1]
```diff
@@ -467,3 +467,52 @@ def merge_dicts(dict1: dict[Any, list[Any]],
...
 
     return dict(merged_dict)
+
+# Copyright 2023 The vLLM team.
+# Adapted from
+# https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/utils.py
+# Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
+from typing import Sequence
+
+import torch
```
Comment (zhuohan123) [inline vllm/utils.py:477]: move the import to the top?

File: vllm/parallel_state.py
Comment (zhuohan123) [inline vllm/parallel_state.py:1]: Can you create a `distributed` folder for `parallel_state.py` and `distributed.py`?

**Review by zhuohan123 — COMMENTED** [review]
LGTM! I feel it's better to not lay all the files flat under `vllm/`. Let's create another directory?

File: vllm/utils.py (hunk 1 above)
Comment (zhuohan123) [reply vllm/utils.py:477]: Or move the utils to `vllm/distributed/utils.py`

Comment (youkaichao) [reply vllm/utils.py:477]: will do linter later.

**Review by zhuohan123 — APPROVED** [review]

**Pull request merged**